### PR TITLE
ADD: .stl file exporting

### PR DESCRIPTION
**Pull request by joeylmaalouf**

### CODE REVIEW

File: src/camera.cpp [hunk 1]
```diff
@@ -18,7 +18,8 @@ void drawCube (tuple3i pos, tuple3i col, int draw_flag) {
 
   if (draw_flag == GL_LINES) {
     glutWireCube(size);
-  } else {
+  }
+  else {
```
Comment (sarahwalters) [inline src/camera.cpp]: :cry: 

(jk)

File: src/fileio.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
+#include <fstream>
+#include <iostream>
+#include <vector>
+#include "fileio.h"
+
+void import_stl (string filepath) {
+}
```
Comment (sarahwalters) [inline src/fileio.cpp]: ooooh this would be cool

File: src/fileio.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+void export_stl (World *world, string filepath) {
+  ofstream fs;
+  fs.open(filepath + ".stl");
```
Comment (sarahwalters) [inline src/fileio.cpp]: d'you have to `touch foo.stl` or similar before running `export_stl`, or does `fs.open` create it for you if it doesn't exist? when does the `!fs.is_open()` case occur?

File: src/fileio.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    cout << "Error: could not open file '" << filepath << ".stl'." << endl;
+    return;
+  }
+
+  fs << "solid " << filepath << endl;
+
+  map<tuple3i, Voxel *> grid = world->getGrid();
+  int n = 0;
+  for (const auto &g : grid) {
+    int size = 1; // same as in drawCube, maybe make into an attribute of Voxel?
```
Comment (sarahwalters) [inline src/fileio.cpp]: :+1: to making into an attribute of Voxel

File: src/fileio.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    };
+
+    vector<vector<vector<float>>> triangles = {
+      {corners[0], corners[1], corners[2]}, {corners[1], corners[2], corners[3]},
+      {corners[0], corners[1], corners[4]}, {corners[1], corners[4], corners[5]},
+      {corners[0], corners[2], corners[4]}, {corners[2], corners[4], corners[6]},
+      {corners[2], corners[3], corners[6]}, {corners[3], corners[6], corners[7]},
+      {corners[4], corners[5], corners[6]}, {corners[5], corners[6], corners[7]},
+      {corners[1], corners[3], corners[5]}, {corners[3], corners[5], corners[7]}
+    };
```
Comment (sarahwalters) [inline src/fileio.cpp]: might be clean to write two little util functions `getCorners(x, y, z, s)` (or even `getCorners(g, s)`) and `getTriangles(corners)`

File: src/fileio.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    for (vector<vector<float>> t : triangles) {
+      fs << "facet normal 0.0e1 0.0e1 " << n << ".0e1" << endl << "    outer loop" << endl;
+      for (int i = 0; i < 3; ++i) {
+        vector<float> point = t[i];
+        fs << "        vertex " << point[0] << "e1 " << point[1] << "e1 " << point[2] << "e1" << endl;
+      }
+      fs << "    endloop" << endl << "endfacet" << endl;
+      ++n;
+    }
```
Comment (sarahwalters) [inline src/fileio.cpp]: i'd be scared to touch this code -- comments about how the string building works, or a pointer to whatever resource you used to figure this out, might be nice

Comment (sarahwalters) [reply src/fileio.cpp]: (immediate questions: do things break if the spaces before vertex and endloop and outer loop aren't there? what are the 0.0e1 and .0e1 markings?)

File: src/fileio.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+void export_stl (World *world, string filepath) {
+  ofstream fs;
+  fs.open(filepath + ".stl");
+  if (!fs.is_open()) {
+    cout << "Error: could not open file '" << filepath << ".stl'." << endl;
+    return;
+  }
+
+  fs << "solid " << filepath << endl;
```
Comment (sarahwalters) [inline src/fileio.cpp]: can this line be down closer to the other stl formatting stuff?

**Review by sarahwalters — CHANGES_REQUESTED** [review]
this is AWESOME.

File: src/fileio.cpp (hunk 3 above)
Comment (joeylmaalouf) [reply src/fileio.cpp]: `fs.open` creates it if it doesn't exist; the `!fs.is_open()` case would only occur if there was some error opening the file for writing, e.g. if another process is currently writing to it or something.

File: src/fileio.cpp (hunk 7 above)
Comment (joeylmaalouf) [reply src/fileio.cpp]: it has to be outside of the loop that write the triangles into the file

File: src/fileio.cpp (hunk 6 above)
Comment (joeylmaalouf) [reply src/fileio.cpp]: https://en.wikipedia.org/wiki/STL_(file_format)#ASCII_STL
I'm not sure if things actually do break without the whitespace and scientific notation, but I figured it's better to follow the official format as closely as possible.

File: src/fileio.cpp (hunk 5 above)
Comment (joeylmaalouf) [reply src/fileio.cpp]: done!

**Review by sarahwalters — APPROVED** [review]

**Pull request merged**